### PR TITLE
GH-117 Land initial websockets and GH-304 Python test

### PR DESCRIPTION
**Pull request by velzevur**
## Messages envelope format
As discussed we will be using the ws for pushing events. We are not going to use it for blocking calls, calls that change the state of the node and etc - for these we'd be using a Swagger generated HTTP API 
### Node sending messages
Broadcasting a new event to all connected websocket clients is done via
```
ws_handler:broadcast(miner, mined_block, [{height, aec_blocks:height(Block)}]).
```
Which results in the following message being send to all ws clients:
```
{"action":"mined_block","origin":"miner","payload":{"height":1}}}
```
This event is currently sent when a new block is mined.

### Client sending messages
Client can send messages through the websocket though the HTTP API would be better suited for this.
If there is no hadneling of the request or there are wrong parameters - there is no response of the request.
Currently there are a couple of requests the server: get header/block by height/hash
Example for getting block with height 1
```
{"action": "get", "target": "chain", "payload": {"type": "block", "height": 1}}
```
Example server's reponse
```
{"action":"requested_data","origin":"chain","payload":{"type":"block","height":1,"block":{"height":1,"nonce":976235507,"pow":[1653605,10024768,10259664,10296433,14584092,18669246,18851946,19040263,19495264,22554524,22813452,25573089,44480753,45966973,47356778,49115947,51174450,51709247,61296826,64386492,64536410,65050643,67241735,68227517,74170775,81261966,82207712,85159905,87140585,89998612,94965092,104072660,104799138,106586019,108240970,109104623,113489287,116937181,117456709,118036804,124884325,127722495],"prev_hash":"u7RpHLzlIZ4qqp+nBVgbR+9BPIla9hcnCd3Xv/H4vUc=","state_hash":"+Gkio7NWSEgKvF+9z1S6bqGQUTVnST5209wUn3lCgdU=","target":553713663,"time":1509476094472,"transactions":[{"data":{"nonce":0,"pubkey":"BBIybDBucGc4MjyXIh/kE7+3/KmNDULPst6WPJbR5cdfzq/I6Zq4QtNf8+9MMVV/MHcPT6cIWznIk455992mAr4="},"signatures":["MEQCIBEP5+bhE2M1ob367qATDuXRJHOhUvPVbWYIjxjaRNq3AiA9geEKD2OcVad9Xik2ekfQZnDGBq7kch1yjnxlmrdqCQ=="],"type":"coinbase"}],"txs_hash":"4rsKBhDvN0nSmv2mtEPUuCgUu0Su4dEFc4J9QrfUB/0=","version":1}}}
```

### Setup
In order to run the tests, you have to prepare the virtenv
```
make python-env
```
There is a default python ws test that relies that you have already `local` node started (on port 3014). It POC and runs indefinetly. You can start it:
```
make python-ws-test
```
it actually runs a python script with default parameters and you can run it by hand with different ones
```
./py/bin/python ./py/tests/ws_client.py --port 3014 --log INFO --handler ws_logic
```
Whrere you can set port, log level (DEBUG, INFO, WARNING, ERROR) and handler. The handler itself is the interesting part and it implements the test's logic. Having the logic of the client separated from the client itself will be usefull for us to have sevral different and independent clients and we could use it for building test applications.
The current python test upgrades its connection to a websocket one, then fetches the Genesis Block and Genesis Header. After that it starts waiting for a new block to be mined. When an event for that is received - it get's the block by its height. Example output:
```
make python-ws-test 
### opened ###
Getting GenesisBlock
<< {"action": "get", "target": "chain", "payload": {"type": "block", "height": 0}}
Getting GenesisHeader
<< {"action": "get", "target": "chain", "payload": {"type": "header", "height": 0}}
>> {"action":"requested_data","origin":"chain","payload":{"type":"block","height":0,"block":{"height":0,"nonce":0,"target":553713663,"time":0,"version":1}}}
Recieved data for block, requested by its height = 0
{u'nonce': 0, u'time': 0, u'target': 553713663, u'version': 1, u'height': 0}
>> {"action":"requested_data","origin":"chain","payload":{"type":"header","height":0,"header":{"height":0,"nonce":0,"target":553713663,"time":0,"version":1}}}
Recieved data for header, requested by its height = 0
{u'nonce': 0, u'time': 0, u'target': 553713663, u'version': 1, u'height': 0}
>> {"action":"mined_block","origin":"miner","payload":{"height":1}}
<< {"action": "get", "target": "chain", "payload": {"type": "block", "height": 1}}
>> {"action":"requested_data","origin":"chain","payload":{"type":"block","height":1,"block":{"height":1,"nonce":976235507,"pow":[1653605,10024768,10259664,10296433,14584092,18669246,18851946,19040263,19495264,22554524,22813452,25573089,44480753,45966973,47356778,49115947,51174450,51709247,61296826,64386492,64536410,65050643,67241735,68227517,74170775,81261966,82207712,85159905,87140585,89998612,94965092,104072660,104799138,106586019,108240970,109104623,113489287,116937181,117456709,118036804,124884325,127722495],"prev_hash":"u7RpHLzlIZ4qqp+nBVgbR+9BPIla9hcnCd3Xv/H4vUc=","state_hash":"+Gkio7NWSEgKvF+9z1S6bqGQUTVnST5209wUn3lCgdU=","target":553713663,"time":1509476094472,"transactions":[{"data":{"nonce":0,"pubkey":"BBIybDBucGc4MjyXIh/kE7+3/KmNDULPst6WPJbR5cdfzq/I6Zq4QtNf8+9MMVV/MHcPT6cIWznIk455992mAr4="},"signatures":["MEQCIBEP5+bhE2M1ob367qATDuXRJHOhUvPVbWYIjxjaRNq3AiA9geEKD2OcVad9Xik2ekfQZnDGBq7kch1yjnxlmrdqCQ=="],"type":"coinbase"}],"txs_hash":"4rsKBhDvN0nSmv2mtEPUuCgUu0Su4dEFc4J9QrfUB/0=","version":1}}}
Recieved data for block, requested by its height = 1
{u'nonce': 976235507, u'target': 553713663, u'transactions': [{u'type': u'coinbase', u'signatures': [u'MEQCIBEP5+bhE2M1ob367qATDuXRJHOhUvPVbWYIjxjaRNq3AiA9geEKD2OcVad9Xik2ekfQZnDGBq7kch1yjnxlmrdqCQ=='], u'data': {u'nonce': 0, u'pubkey': u'BBIybDBucGc4MjyXIh/kE7+3/KmNDULPst6WPJbR5cdfzq/I6Zq4QtNf8+9MMVV/MHcPT6cIWznIk455992mAr4='}}], u'pow': [1653605, 10024768, 10259664, 10296433, 14584092, 18669246, 18851946, 19040263, 19495264, 22554524, 22813452, 25573089, 44480753, 45966973, 47356778, 49115947, 51174450, 51709247, 61296826, 64386492, 64536410, 65050643, 67241735, 68227517, 74170775, 81261966, 82207712, 85159905, 87140585, 89998612, 94965092, 104072660, 104799138, 106586019, 108240970, 109104623, 113489287, 116937181, 117456709, 118036804, 124884325, 127722495], u'time': 1509476094472, u'height': 1, u'version': 1, u'txs_hash': u'4rsKBhDvN0nSmv2mtEPUuCgUu0Su4dEFc4J9QrfUB/0=', u'prev_hash': u'u7RpHLzlIZ4qqp+nBVgbR+9BPIla9hcnCd3Xv/H4vUc=', u'state_hash': u'+Gkio7NWSEgKvF+9z1S6bqGQUTVnST5209wUn3lCgdU='}



```

### CODE REVIEW

**Comment by sennui** [comment]
Websocket API will be sort of equivalent of HTTP api for local app interface (trusted).
Are we going to make an api available for both protocols?
I would like to make it possible. In such scenerio, should "action" hold some params that used to be in URL or headers?

File: apps/aehttp/src/ws_int_dispatch.erl [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    end.
+
+is_valid(target, invalid_message, _SomePayload) ->
+    {false, <<"some reason">>};
+is_valid(_, _, _) ->
+    true.
+
+do_execute(silence_dialyzer, _, _) ->
+    {error, <<"oh, no">>};
+do_execute(mining, stop, _) ->
```
Comment (sennui) [inline apps/aehttp/src/ws_int_dispatch.erl]: These are all mocks of the functionality? Do we need them, given we move away from "actions" via WS?

File: apps/aehttp/src/ws_handler.erl [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        case Payload of
+            _ when is_list(Payload) ->
+                length(Payload) == 0;
+            _ when is_map(Payload) ->
+                maps:size(Payload) == 0
+        end,
+    Msg =
+        case EmptyPayload of
+            true ->
+                MsgWithoutPayload;
```
Comment (sennui) [inline apps/aehttp/src/ws_handler.erl]: what's an example of the message without payload?

File: apps/aehttp/src/ws_task_worker.erl [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    WsPid = self(),
+    spawn(fun() -> ws_task_worker_sup:spawn_child_for_msg(Msg, WsPid) end).
+
+start_link(Args) ->
+    gen_server:start_link(?MODULE, Args, []).
+
+init([ws_task_worker_args]) ->
+    {ok, #state{}}.
+
+handle_call({handle_msg, {MsgBin, WsPid}}, _From, #state{}=State) ->
```
Comment (sennui) [inline apps/aehttp/src/ws_task_worker.erl]: what actions do we support?

File: apps/aehttp/src/ws_handler.erl (hunk 2 above)
Comment (velzevur) [reply apps/aehttp/src/ws_handler.erl]: Well we could have. I don't know if we're to send a message `the miner has stopped` but we could send a message without a payload.

**Pull request merged**